File: android/openmekit/src/main/kotlin/org/merlos/openmekit/ClientConfigParser.kt
package org.merlos.openmekit

import org.json.JSONObject

/**
 * Parses and serialises the openme client configuration format.
 *
 * Supports two input formats:
 * - **YAML** — the `config.yaml` file produced by `openme add` on the server.
 * - **JSON** — the QR code payload produced by `openme qr`.
 *
 * The YAML parser is a lightweight, line-based reader that handles the specific subset of YAML
 * produced by `openme add`. It auto-detects indent width (2-space / 4-space) and works with
 * both go-yaml v2 and go-yaml v3 output.
 *
 * ### Expected YAML schema
 * ```yaml
 * profiles:
 *   <name>:
 *     server_host: "..."
 *     server_udp_port: 54154
 *     server_pubkey: "..."
 *     private_key: "..."
 *     public_key: "..."
 *     post_knock: "..."
 * ```
 *
 * ### QR JSON schema
 * ```json
 * {
 *   "profile": "my-server",
 *   "host": "203.0.113.1",
 *   "udp_port": 54154,
 *   "server_pubkey": "<base64>",
 *   "client_privkey": "<base64>",
 *   "client_pubkey": "<base64>"
 * }
 * ```
 *
 * For the full config format reference see the
 * [Client Configuration](https://openme.merlos.org/docs/configuration/client.html) docs.
 */
object ClientConfigParser {

    // ─── YAML ───────────────────────────────────────────────────────────────────

    /**
     * Parses a `config.yaml` YAML string into a map of [Profile] values keyed by name.
     *
     * The parser auto-detects indent width and handles both 2-space (go-yaml v2) and
     * 4-space (go-yaml v3) indentation. Malformed or incomplete profiles are silently
     * skipped so a partially-valid config still loads the good entries.
     *
     * @param yaml Full contents of a `config.yaml` file.
     * @return Map of profile name → [Profile]. Empty if no valid profiles are found.
     * @throws ParserError.NoProfilesFound if the YAML contains no `profiles:` section
     *   or every profile fails validation.
     */
    @Throws(ParserError::class)
    fun parseYaml(yaml: String): Map<String, Profile> {
        val profiles = mutableMapOf<String, Profile>()
        val rawLines = yaml.lines()

        var inProfiles = false
        var currentName: String? = null
        var currentDict = mutableMapOf<String, String>()
        var profileIndent: Int? = null
        var kvIndent: Int? = null

        fun leadingSpaces(s: String): Int {
            var count = 0
            for (ch in s) {
                when (ch) {
                    ' ' -> count++
                    '\t' -> count += 4
                    else -> break
                }
            }
            return count
        }

        fun parseValue(raw: String): String {
            val trimmed = raw.trim()
            if (trimmed.isEmpty()) return ""
            val quoteChars = setOf('"', '\'')
            if (trimmed.first() in quoteChars) {
                val openQuote = trimmed.first()
                val rest = trimmed.drop(1)
                val closeIdx = rest.indexOf(openQuote)
                if (closeIdx >= 0) return rest.substring(0, closeIdx)
            }
            // Unquoted: strip inline comment
            val commentIdx = trimmed.indexOf(" #")
            return if (commentIdx >= 0) trimmed.substring(0, commentIdx).trim() else trimmed
        }

        fun commitProfile(name: String, dict: Map<String, String>) {
            val host = dict["server_host"] ?: return
            val port = dict["server_udp_port"]?.toIntOrNull() ?: 54154
            val serverPub = dict["server_pubkey"]?.takeIf { it.isNotBlank() } ?: return
            val privKey = dict["private_key"]?.takeIf { it.isNotBlank() } ?: return
            val pubKey = dict["public_key"] ?: ""
            val postKnock = dict["post_knock"] ?: ""
            profiles[name] = Profile(
                name = name,
                serverHost = host,
                serverUDPPort = port,
                serverPubKey = serverPub,
                privateKey = privKey,
                publicKey = pubKey,
                postKnock = postKnock,
            )
        }

        for (rawLine in rawLines) {
            // Skip blank, comment, and document-marker lines
            val stripped = rawLine.trimEnd()
            if (stripped.isBlank()) continue
            val trimmed = stripped.trim()
            if (trimmed.startsWith("#") || trimmed == "---" || trimmed == "..." || trimmed.startsWith("%")) continue

            val indent = leadingSpaces(stripped)

            // Looking for the root `profiles:` key (must be at indent 0)
            if (!inProfiles) {
                if (indent == 0 && trimmed == "profiles:") {
                    inProfiles = true
                }
                continue
            }

            // We are inside the profiles: block
            val colonIdx = trimmed.indexOf(':')
            if (colonIdx < 0) continue
            val key = trimmed.substring(0, colonIdx).trim()
            val valueRaw = trimmed.substring(colonIdx + 1)

            when {
                // Profile-name line — deduce profileIndent on first encounter
                profileIndent == null || indent == profileIndent -> {
                    // Save previous profile
                    currentName?.let { commitProfile(it, currentDict) }
                    profileIndent = indent
                    currentName = key
                    currentDict = mutableMapOf()
                    kvIndent = null
                }
                // Key-value line — deduce kvIndent on first encounter
                kvIndent == null || indent == kvIndent -> {
                    kvIndent = indent
                    currentDict[key] = parseValue(valueRaw)
                }
                // Deeper nesting or unexpected indent — ignore
            }
        }

        currentName?.let { commitProfile(it, currentDict) }

        if (profiles.isEmpty()) throw ParserError.NoProfilesFound
        return profiles
    }

    /**
     * Serialises a map of profiles to a `config.yaml` string compatible with
     * the openme CLI (go-yaml v3, 4-space indent).
     *
     * @param profiles Map of profile name → [Profile].
     * @return YAML string ready to be written to `config.yaml`.
     */
    fun toYaml(profiles: Map<String, Profile>): String {
        val sb = StringBuilder("profiles:\n")
        for ((_, p) in profiles) {
            sb.append("    ${p.name}:\n")
            sb.append("        server_host: \"${p.serverHost}\"\n")
            sb.append("        server_udp_port: ${p.serverUDPPort}\n")
            sb.append("        server_pubkey: \"${p.serverPubKey}\"\n")
            sb.append("        private_key: \"${p.privateKey}\"\n")
            sb.append("        public_key: \"${p.publicKey}\"\n")
            sb.append("        post_knock: \"${p.postKnock}\"\n")
        }
        return sb.toString()
    }

    // ─── QR JSON ────────────────────────────────────────────────────────────────

    /**
     * Parses the JSON payload from an openme QR code into a [Profile].
     *
     * QR codes are generated by `openme qr <profile-name>` on the server.
     * They encode a JSON object — **not** the YAML config format.
     *
     * Expected JSON fields:
     * - `"profile"` — profile name
     * - `"host"` — server hostname or IP
     * - `"udp_port"` — server UDP port
     * - `"server_pubkey"` — Base64-encoded server Curve25519 public key
     * - `"client_privkey"` — Base64-encoded client Ed25519 private key
     * - `"client_pubkey"` — Base64-encoded client Ed25519 public key
     *
     * @param json Raw JSON string decoded from a QR code.
     * @return A fully populated [Profile].
     * @throws ParserError.InvalidQRPayload if required fields are missing or the
     *   string is not valid JSON.
     */
    @Throws(ParserError::class)
    fun parseQRPayload(json: String): Profile {
        return try {
            val obj = JSONObject(json)
            val name = obj.optString("profile").takeIf { it.isNotBlank() }
                ?: throw ParserError.InvalidQRPayload
            val host = obj.optString("host").takeIf { it.isNotBlank() }
                ?: throw ParserError.InvalidQRPayload
            val port = obj.optInt("udp_port", 54154)
            val serverPub = obj.optString("server_pubkey").takeIf { it.isNotBlank() }
                ?: throw ParserError.InvalidQRPayload
            val privKey = obj.optString("client_privkey").takeIf { it.isNotBlank() } ?: ""
            val pubKey = obj.optString("client_pubkey") ?: ""
            Profile(
                name = name,
                serverHost = host,
                serverUDPPort = port,
                serverPubKey = serverPub,
                privateKey = privKey,
                publicKey = pubKey,
            )
        } catch (e: ParserError) {
            throw e
        } catch (e: Exception) {
            throw ParserError.InvalidQRPayload
        }
    }
}

/**
 * Errors thrown by [ClientConfigParser].
 */
sealed class ParserError : Exception() {
    /** The YAML contains no `profiles:` section or every profile fails validation. */
    object NoProfilesFound : ParserError() {
        override val message = "No valid profiles found in the configuration."
        private fun readResolve(): Any = NoProfilesFound
    }

    /** The QR JSON payload is missing required fields or is not valid JSON. */
    object InvalidQRPayload : ParserError() {
        override val message = "Invalid QR payload: required fields are missing or the JSON is malformed."
        private fun readResolve(): Any = InvalidQRPayload
    }
}

File: windows/OpenMeKit/ClientConfigParser.cs
using System.Text.RegularExpressions;

namespace OpenMeKit;

/// <summary>
/// Parses and serialises the <c>~/.openme/config.yaml</c> client configuration file.
/// </summary>
/// <remarks>
/// The parser is a lightweight, line-based YAML reader that handles the specific subset
/// of YAML produced by <c>openme add</c>. It supports both the 2-space indentation of
/// go-yaml v2 and the 4-space indentation of go-yaml v3, as well as the QR-code JSON
/// payload format used by the mobile apps.
///
/// Expected YAML schema:
/// <code>
/// profiles:
///   &lt;name&gt;:
///     server_host: "..."
///     server_udp_port: 54154
///     server_pubkey: "..."
///     private_key: "..."
///     public_key:  "..."
///     post_knock:  "..."
/// </code>
/// </remarks>
public static class ClientConfigParser
{
    // ── YAML parsing ──────────────────────────────────────────────────────────

    /// <summary>
    /// Parses a <c>config.yaml</c> string into a dictionary of <see cref="Profile"/> objects
    /// keyed by profile name.
    /// </summary>
    /// <param name="yaml">Content of a config.yaml produced by <c>openme add</c>.</param>
    /// <returns>Dictionary mapping profile name → <see cref="Profile"/>.</returns>
    /// <exception cref="ClientConfigException">Thrown when no profiles can be found.</exception>
    public static Dictionary<string, Profile> ParseYaml(string yaml)
    {
        var profiles = new Dictionary<string, Profile>(StringComparer.Ordinal);
        var lines = yaml.ReplaceLineEndings("\n").Split('\n');

        // Detect indentation level (2 or 4 spaces) from the first profile-name line
        // that appears directly under "profiles:".
        int indent = 0;
        bool inProfiles = false;

        foreach (var raw in lines)
        {
            if (raw.TrimStart().StartsWith("profiles:", StringComparison.Ordinal))
            {
                inProfiles = true;
                continue;
            }
            if (!inProfiles) continue;

            // First non-empty line after "profiles:" is a profile name at indent N.
            var trimmed = raw.TrimEnd();
            if (trimmed.Length == 0) continue;

            int spaces = trimmed.Length - trimmed.TrimStart().Length;
            if (spaces > 0) { indent = spaces; break; }
        }

        if (indent == 0) indent = 2; // fallback

        Profile? current = null;
        inProfiles = false;

        foreach (var raw in lines)
        {
            var line = raw.TrimEnd();
            if (line.TrimStart().StartsWith("profiles:", StringComparison.Ordinal))
            {
                inProfiles = true;
                continue;
            }
            if (!inProfiles) continue;
            if (line.Trim().Length == 0) continue;

            int depth = (line.Length - line.TrimStart().Length) / indent;

            if (depth == 1)
            {
                // Profile name line: "  alice:"
                if (current != null && !string.IsNullOrEmpty(current.Name))
                    profiles[current.Name] = current;

                var name = line.Trim().TrimEnd(':');
                current = new Profile { Name = name };
                continue;
            }

            if (depth == 2 && current != null)
            {
                var (key, value) = SplitKeyValue(line.Trim());
                switch (key)
                {
                    case "server_host":     current.ServerHost    = value; break;
                    case "server_udp_port": current.ServerUdpPort = ushort.TryParse(value, out var p) ? p : (ushort)54154; break;
                    case "server_pubkey":   current.ServerPubKey  = value; break;
                    case "private_key":     current.PrivateKey    = value; break;
                    case "public_key":      current.PublicKey     = value; break;
                    case "post_knock":      current.PostKnock     = value; break;
                }
            }
        }

        if (current != null && !string.IsNullOrEmpty(current.Name))
            profiles[current.Name] = current;

        if (profiles.Count == 0)
            throw new ClientConfigException(ClientConfigError.NoProfilesFound,
                "No profiles found in the YAML. Make sure the text starts with 'profiles:'.");

        return profiles;
    }

    // ── QR payload parsing ────────────────────────────────────────────────────

    /// <summary>
    /// Parses the JSON payload encoded in an openme QR code.
    /// </summary>
    /// <param name="json">
    /// JSON string with keys: <c>profile</c>, <c>host</c>, <c>udp_port</c>,
    /// <c>server_pubkey</c>, <c>client_privkey</c>, <c>client_pubkey</c>.
    /// </param>
    /// <returns>A populated <see cref="Profile"/>.</returns>
    /// <exception cref="ClientConfigException">Thrown when required fields are missing.</exception>
    public static Profile ParseQrPayload(string json)
    {
        // Lightweight JSON field extraction without a full JSON parser dependency.
        string? Get(string field)
        {
            // Matches: "field": "value"  or  "field":"value"
            var m = Regex.Match(json, $"\"{Regex.Escape(field)}\"\\s*:\\s*\"([^\"]*)\"");
            return m.Success ? m.Groups[1].Value : null;
        }
        string? GetInt(string field)
        {
            var m = Regex.Match(json, $"\"{Regex.Escape(field)}\"\\s*:\\s*(\\d+)");
            return m.Success ? m.Groups[1].Value : null;
        }

        var name   = Get("profile");
        var host   = Get("host");
        var port   = GetInt("udp_port");
        var sPub   = Get("server_pubkey");
        var cPriv  = Get("client_privkey");
        var cPub   = Get("client_pubkey");

        if (string.IsNullOrEmpty(name) || string.IsNullOrEmpty(host) ||
            string.IsNullOrEmpty(sPub)  || string.IsNullOrEmpty(cPriv))
            throw new ClientConfigException(ClientConfigError.InvalidQrPayload,
                "QR payload is missing required fields (profile, host, server_pubkey, client_privkey).");

        return new Profile
        {
            Name          = name,
            ServerHost    = host,
            ServerUdpPort = ushort.TryParse(port, out var p) ? p : (ushort)54154,
            ServerPubKey  = sPub,
            PrivateKey    = cPriv,
            PublicKey     = cPub ?? string.Empty,
        };
    }

    // ── YAML serialisation ────────────────────────────────────────────────────

    /// <summary>
    /// Serialises a collection of profiles to the <c>config.yaml</c> format compatible
    /// with the openme CLI.
    /// </summary>
    public static string ToYaml(IEnumerable<Profile> profiles)
    {
        var sb = new System.Text.StringBuilder();
        sb.AppendLine("profiles:");
        foreach (var p in profiles)
        {
            sb.AppendLine($"    {p.Name}:");
            sb.AppendLine($"        server_host: {p.ServerHost}");
            sb.AppendLine($"        server_udp_port: {p.ServerUdpPort}");
            sb.AppendLine($"        server_pubkey: {p.ServerPubKey}");
            sb.AppendLine($"        private_key: {p.PrivateKey}");
            sb.AppendLine($"        public_key: {p.PublicKey}");
            if (!string.IsNullOrEmpty(p.PostKnock))
                sb.AppendLine($"        post_knock: {p.PostKnock}");
        }
        return sb.ToString();
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private static (string key, string value) SplitKeyValue(string line)
    {
        var idx = line.IndexOf(':', StringComparison.Ordinal);
        if (idx < 0) return (line, string.Empty);
        var key   = line[..idx].Trim();
        var value = line[(idx + 1)..].Trim().Trim('"');
        return (key, value);
    }
}

/// <summary>Error categories for <see cref="ClientConfigException"/>.</summary>
public enum ClientConfigError
{
    /// <summary>The YAML contained no <c>profiles:</c> section with entries.</summary>
    NoProfilesFound,
    /// <summary>The QR-code JSON payload was missing required fields or was malformed.</summary>
    InvalidQrPayload,
}

/// <summary>Exception thrown by <see cref="ClientConfigParser"/>.</summary>
public sealed class ClientConfigException(ClientConfigError kind, string message)
    : Exception(message)
{
    /// <summary>The specific parse error that occurred.</summary>
    public ClientConfigError Kind { get; } = kind;
}
